### PR TITLE
Read and notify firmware version

### PR DESCRIPTION
**Pull request by auras**

### CODE REVIEW

File: nuimo/src/main/kotlin/com/senic/nuimo/NuimoBluetoothController.kt [hunk 1]
```diff
@@ -62,12 +62,23 @@ class NuimoBluetoothController(bluetoothDevice: BluetoothDevice, context: Contex
         matrixWriter = null
     }
 
+    private fun readCharacteristics() {
+        readBatteryPercentage()
+        readFirmwareVersion()
+    }
+
     private fun readBatteryPercentage() {
```
Comment (larsblumberg) [inline nuimo/src/main/kotlin/com/senic/nuimo/NuimoBluetoothController.kt]: Please give this method arguments to make it reusable for both characteristics

File: nuimo/src/main/kotlin/com/senic/nuimo/NuimoBluetoothController.kt [hunk 2]
```diff
@@ -62,12 +62,23 @@ class NuimoBluetoothController(bluetoothDevice: BluetoothDevice, context: Contex
         matrixWriter = null
     }
 
+    private fun readCharacteristics() {
+        readBatteryPercentage()
+        readFirmwareVersion()
+    }
+
```
Comment (larsblumberg) [inline nuimo/src/main/kotlin/com/senic/nuimo/NuimoBluetoothController.kt]: Please call both read methods directly from `onCharacteristicChanged`, no need for that wrapper method

File: nuimo/src/main/kotlin/com/senic/nuimo/NuimoBluetoothController.kt [hunk 3]
```diff
@@ -125,11 +136,16 @@ class NuimoBluetoothController(bluetoothDevice: BluetoothDevice, context: Contex
         }
 
         override fun onCharacteristicRead(gatt: BluetoothGatt, characteristic: BluetoothGattCharacteristic, status: Int) {
+            writeQueue.next()
+
             if (characteristic.uuid.equals(BATTERY_CHARACTERISTIC_UUID)) {
-                matrixWriter?.onWrite()
```
Comment (larsblumberg) [inline nuimo/src/main/kotlin/com/senic/nuimo/NuimoBluetoothController.kt:129]: What's the reason for this change?

File: nuimo/src/main/kotlin/com/senic/nuimo/NuimoBluetoothController.kt [hunk 4]
```diff
@@ -125,11 +136,16 @@ class NuimoBluetoothController(bluetoothDevice: BluetoothDevice, context: Contex
...
                 val batteryPercentage = characteristic.getIntValue(BluetoothGattCharacteristic.FORMAT_UINT8, 0) ?: -1
                 notifyListeners { it.onBatteryPercentageChange(batteryPercentage) }
             }
+            else if (characteristic.uuid.equals(FIRMWARE_VERSION_CHARACTERISTIC_UUID)) {
+                val firmwareVersion = characteristic.getStringValue(0)
```
Comment (larsblumberg) [inline nuimo/src/main/kotlin/com/senic/nuimo/NuimoBluetoothController.kt]: Don't use `!!` but test for `null`. Don't notify if it's `null`

Comment (auras) [inline nuimo/src/main/kotlin/com/senic/nuimo/NuimoBluetoothController.kt]: That's why asked you today if it can be null

Comment (larsblumberg) [inline nuimo/src/main/kotlin/com/senic/nuimo/NuimoBluetoothController.kt]: I understood that you were asking whether the firmware characteristic is there at all.

Don't rely on Android that it actually returns a non-null value – or is it documented to always return a non-null value?

File: nuimo/src/main/kotlin/com/senic/nuimo/NuimoBluetoothController.kt (hunk 3 above)
Comment (auras) [inline nuimo/src/main/kotlin/com/senic/nuimo/NuimoBluetoothController.kt:129]: 1) It needs to happen for all cases, not just battery characteristic
2) It wasn't the proper method to call here. writeQueue.next() should have been called. It only worked because onWrite
() also called writeQueue.next()

File: nuimo/src/main/kotlin/com/senic/nuimo/NuimoBluetoothController.kt (hunk 4 above)
Comment (auras) [inline nuimo/src/main/kotlin/com/senic/nuimo/NuimoBluetoothController.kt]: If it's always there why should it return null?

Comment (larsblumberg) [inline nuimo/src/main/kotlin/com/senic/nuimo/NuimoBluetoothController.kt]: Characteristic is present != Android returns a non-null value. Please test for null, never force un-wrap an optional unless there's a good reason.

Comment (auras) [inline nuimo/src/main/kotlin/com/senic/nuimo/NuimoBluetoothController.kt]: ok, fixed it

File: nuimo/src/main/kotlin/com/senic/nuimo/NuimoBluetoothController.kt [hunk 5]
```diff
@@ -125,11 +124,16 @@ class NuimoBluetoothController(bluetoothDevice: BluetoothDevice, context: Contex
...
+                if (firmwareVersion != null) notifyListeners { it.onFirmwareVersionRead(firmwareVersion) }
+            }
```
Comment (larsblumberg) [inline nuimo/src/main/kotlin/com/senic/nuimo/NuimoBluetoothController.kt:138]: One last thing, please use

```
when (characteristic.uuid)
    BATTERY_CHARACTERISTIC_UUID -> { ...
    ... ->
```

Comment (auras) [inline nuimo/src/main/kotlin/com/senic/nuimo/NuimoBluetoothController.kt:138]: done
